File: crates/stateset-embedded/src/accounts_receivable.rs
//! Accounts Receivable operations
//!
//! Comprehensive accounts receivable management supporting:
//! - AR aging reports and analysis
//! - Collection activity tracking
//! - Credit memos and applications
//! - Write-offs
//! - Payment applications
//! - Customer statements
//!
//! # Example
//!
//! ```rust,ignore
//! use stateset_embedded::Commerce;
//!
//! let commerce = Commerce::new("./store.db")?;
//!
//! // Get AR aging summary
//! let aging = commerce.accounts_receivable().get_aging_summary()?;
//! println!("Total outstanding: ${}", aging.total);
//! println!("Current: ${}", aging.current);
//! println!("30+ days: ${}", aging.days_1_30);
//! # Ok::<(), stateset_embedded::CommerceError>(())
//! ```

use rust_decimal::Decimal;
use stateset_core::{
    ApplyCreditMemo, ApplyPaymentToInvoices, ArAgingFilter, ArAgingSummary, ArPaymentApplication,
    CollectionActivity, CollectionActivityFilter, CollectionStatus, CreateCollectionActivity,
    CreateCreditMemo, CreateWriteOff, CreditMemo, CreditMemoFilter, CustomerArAging,
    CustomerArSummary, CustomerStatement, DunningLetterType, GenerateStatementRequest, Invoice,
    Result, WriteOff, WriteOffFilter,
};
use stateset_db::Database;
use std::sync::Arc;
use uuid::Uuid;

/// Accounts Receivable interface.
pub struct AccountsReceivable {
    db: Arc<dyn Database>,
}

impl std::fmt::Debug for AccountsReceivable {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("AccountsReceivable").finish_non_exhaustive()
    }
}

impl AccountsReceivable {
    pub(crate) fn new(db: Arc<dyn Database>) -> Self {
        Self { db }
    }

    // ========================================================================
    // Aging Reports
    // ========================================================================

    /// Get overall AR aging summary.
    ///
    /// # Example
    ///
    /// ```rust,ignore
    /// use stateset_embedded::Commerce;
    ///
    /// let commerce = Commerce::new(":memory:")?;
    ///
    /// let aging = commerce.accounts_receivable().get_aging_summary()?;
    /// println!("Current: ${}", aging.current);
    /// println!("1-30 days: ${}", aging.days_1_30);
    /// println!("31-60 days: ${}", aging.days_31_60);
    /// println!("61-90 days: ${}", aging.days_61_90);
    /// println!("90+ days: ${}", aging.days_over_90);
    /// println!("Total: ${}", aging.total);
    /// # Ok::<(), stateset_embedded::CommerceError>(())
    /// ```
    pub fn get_aging_summary(&self) -> Result<ArAgingSummary> {
        self.db.accounts_receivable().get_aging_summary()
    }

    /// Get AR aging for a specific customer.
    pub fn get_customer_aging(&self, customer_id: Uuid) -> Result<Option<CustomerArAging>> {
        self.db.accounts_receivable().get_customer_aging(customer_id)
    }

    /// Get detailed AR aging report with filtering.
    pub fn get_aging_report(&self, filter: ArAgingFilter) -> Result<Vec<CustomerArAging>> {
        self.db.accounts_receivable().get_aging_report(filter)
    }

    // ========================================================================
    // Collection Activities
    // ========================================================================

    /// Log a collection activity (call, email, dunning letter, etc.).
    ///
    /// # Example
    ///
    /// ```rust,ignore
    /// use stateset_embedded::{Commerce, CreateCollectionActivity, CollectionActivityType};
    /// use uuid::Uuid;
    ///
    /// let commerce = Commerce::new(":memory:")?;
    ///
    /// let activity = commerce.accounts_receivable().log_collection_activity(
    ///     CreateCollectionActivity {
    ///         invoice_id: Uuid::new_v4(),
    ///         activity_type: CollectionActivityType::PhoneCall,
    ///         notes: Some("Spoke with customer, promised to pay by Friday".into()),
    ///         contact_method: Some("Phone".into()),
    ///         contact_result: Some("Promise to pay".into()),
    ///         performed_by: Some("John Collector".into()),
    ///         ..Default::default()
    ///     }
    /// )?;
    /// # Ok::<(), stateset_embedded::CommerceError>(())
    /// ```
    pub fn log_collection_activity(
        &self,
        input: CreateCollectionActivity,
    ) -> Result<CollectionActivity> {
        self.db.accounts_receivable().log_collection_activity(input)
    }

    /// List collection activities with filtering.
    pub fn list_collection_activities(
        &self,
        filter: CollectionActivityFilter,
    ) -> Result<Vec<CollectionActivity>> {
        self.db.accounts_receivable().list_collection_activities(filter)
    }

    /// Update the collection status of an invoice.
    pub fn update_collection_status(
        &self,
        invoice_id: Uuid,
        status: CollectionStatus,
    ) -> Result<()> {
        self.db.accounts_receivable().update_collection_status(invoice_id.into(), status)
    }

    // ========================================================================
    // Dunning
    // ========================================================================

    /// Get invoices that are due for dunning letters.
    pub fn get_invoices_due_for_dunning(&self) -> Result<Vec<Invoice>> {
        self.db.accounts_receivable().get_invoices_due_for_dunning()
    }

    /// Send a dunning letter and log the activity.
    ///
    /// # Example
    ///
    /// ```rust,ignore
    /// use stateset_embedded::{Commerce, DunningLetterType};
    /// use uuid::Uuid;
    ///
    /// let commerce = Commerce::new(":memory:")?;
    ///
    /// let activity = commerce.accounts_receivable().send_dunning_letter(
    ///     Uuid::new_v4(), // invoice_id
    ///     DunningLetterType::Reminder1,
    ///     Some("ar_system"),
    /// )?;
    /// # Ok::<(), stateset_embedded::CommerceError>(())
    /// ```
    pub fn send_dunning_letter(
        &self,
        invoice_id: Uuid,
        letter_type: DunningLetterType,
        sent_by: Option<&str>,
    ) -> Result<CollectionActivity> {
        self.db.accounts_receivable().send_dunning_letter(invoice_id.into(), letter_type, sent_by)
    }

    // ========================================================================
    // Write-offs
    // ========================================================================

    /// Create a write-off for an uncollectable invoice.
    ///
    /// # Example
    ///
    /// ```rust,ignore
    /// use stateset_embedded::{Commerce, CreateWriteOff, WriteOffReason};
    /// use rust_decimal_macros::dec;
    /// use uuid::Uuid;
    ///
    /// let commerce = Commerce::new(":memory:")?;
    ///
    /// let write_off = commerce.accounts_receivable().create_write_off(CreateWriteOff {
    ///     invoice_id: Uuid::new_v4(),
    ///     amount: dec!(500.00),
    ///     reason: WriteOffReason::Uncollectable,
    ///     notes: Some("Customer bankruptcy".into()),
    ///     approved_by: Some("CFO".into()),
    /// })?;
    /// # Ok::<(), stateset_embedded::CommerceError>(())
    /// ```
    pub fn create_write_off(&self, input: CreateWriteOff) -> Result<WriteOff> {
        self.db.accounts_receivable().create_write_off(input)
    }

    /// Get a write-off by ID.
    pub fn get_write_off(&self, id: Uuid) -> Result<Option<WriteOff>> {
        self.db.accounts_receivable().get_write_off(id)
    }

    /// List write-offs with filtering.
    pub fn list_write_offs(&self, filter: WriteOffFilter) -> Result<Vec<WriteOff>> {
        self.db.accounts_receivable().list_write_offs(filter)
    }

    /// Reverse a write-off (restore invoice to collections).
    pub fn reverse_write_off(&self, id: Uuid) -> Result<WriteOff> {
        self.db.accounts_receivable().reverse_write_off(id)
    }

    // ========================================================================
    // Credit Memos
    // ========================================================================

    /// Create a credit memo for a customer.
    ///
    /// # Example
    ///
    /// ```rust,ignore
    /// use stateset_embedded::{Commerce, CreateCreditMemo, CreditMemoReason};
    /// use rust_decimal_macros::dec;
    /// use uuid::Uuid;
    ///
    /// let commerce = Commerce::new(":memory:")?;
    ///
    /// let memo = commerce.accounts_receivable().create_credit_memo(CreateCreditMemo {
    ///     customer_id: Uuid::new_v4(),
    ///     original_invoice_id: Some(Uuid::new_v4()),
    ///     reason: CreditMemoReason::ReturnCredit,
    ///     amount: dec!(150.00),
    ///     notes: Some("Credit for returned merchandise".into()),
    /// })?;
    /// # Ok::<(), stateset_embedded::CommerceError>(())
    /// ```
    pub fn create_credit_memo(&self, input: CreateCreditMemo) -> Result<CreditMemo> {
        self.db.accounts_receivable().create_credit_memo(input)
    }

    /// Get a credit memo by ID.
    pub fn get_credit_memo(&self, id: Uuid) -> Result<Option<CreditMemo>> {
        self.db.accounts_receivable().get_credit_memo(id)
    }

    /// Get a credit memo by number.
    pub fn get_credit_memo_by_number(&self, number: &str) -> Result<Option<CreditMemo>> {
        self.db.accounts_receivable().get_credit_memo_by_number(number)
    }

    /// List credit memos with filtering.
    pub fn list_credit_memos(&self, filter: CreditMemoFilter) -> Result<Vec<CreditMemo>> {
        self.db.accounts_receivable().list_credit_memos(filter)
    }

    /// Apply a credit memo to an invoice.
    pub fn apply_credit_memo(&self, input: ApplyCreditMemo) -> Result<CreditMemo> {
        self.db.accounts_receivable().apply_credit_memo(input)
    }

    /// Void a credit memo (only if not yet applied).
    pub fn void_credit_memo(&self, id: Uuid) -> Result<CreditMemo> {
        self.db.accounts_receivable().void_credit_memo(id)
    }

    /// Get unapplied credit memos for a customer.
    pub fn get_unapplied_credits(&self, customer_id: Uuid) -> Result<Vec<CreditMemo>> {
        self.db.accounts_receivable().get_unapplied_credits(customer_id)
    }

    // ========================================================================
    // Payment Applications
    // ========================================================================

    /// Apply a payment to one or more invoices.
    ///
    /// # Example
    ///
    /// ```rust,ignore
    /// use stateset_embedded::{Commerce, ApplyPaymentToInvoices, InvoicePaymentApplication};
    /// use rust_decimal_macros::dec;
    /// use uuid::Uuid;
    ///
    /// let commerce = Commerce::new(":memory:")?;
    ///
    /// let applications = commerce.accounts_receivable().apply_payment_to_invoices(
    ///     ApplyPaymentToInvoices {
    ///         payment_id: Uuid::new_v4(),
    ///         applications: vec![
    ///             InvoicePaymentApplication {
    ///                 invoice_id: Uuid::new_v4(),
    ///                 amount: dec!(500.00),
    ///             },
    ///             InvoicePaymentApplication {
    ///                 invoice_id: Uuid::new_v4(),
    ///                 amount: dec!(250.00),
    ///             },
    ///         ],
    ///     }
    /// )?;
    /// # Ok::<(), stateset_embedded::CommerceError>(())
    /// ```
    pub fn apply_payment_to_invoices(
        &self,
        input: ApplyPaymentToInvoices,
    ) -> Result<Vec<ArPaymentApplication>> {
        self.db.accounts_receivable().apply_payment_to_invoices(input)
    }

    /// Get all payment applications for a payment.
    pub fn get_payment_applications(&self, payment_id: Uuid) -> Result<Vec<ArPaymentApplication>> {
        self.db.accounts_receivable().get_payment_applications(payment_id)
    }

    /// Unapply a payment application (remove application, restore invoice balance).
    pub fn unapply_payment(&self, application_id: Uuid) -> Result<()> {
        self.db.accounts_receivable().unapply_payment(application_id)
    }

    // ========================================================================
    // Customer Summary & Statements
    // ========================================================================

    /// Get AR summary for a customer.
    pub fn get_customer_summary(&self, customer_id: Uuid) -> Result<Option<CustomerArSummary>> {
        self.db.accounts_receivable().get_customer_summary(customer_id)
    }

    /// Generate a customer statement.
    ///
    /// # Example
    ///
    /// ```rust,ignore
    /// use stateset_embedded::{Commerce, GenerateStatementRequest};
    /// use chrono::Utc;
    /// use uuid::Uuid;
    ///
    /// let commerce = Commerce::new(":memory:")?;
    ///
    /// let statement = commerce.accounts_receivable().generate_statement(
    ///     GenerateStatementRequest {
    ///         customer_id: Uuid::new_v4(),
    ///         period_start: None, // defaults to 30 days ago
    ///         period_end: None,   // defaults to now
    ///         include_paid_invoices: Some(false),
    ///     }
    /// )?;
    ///
    /// println!("Statement for: {}", statement.customer_name);
    /// println!("Closing balance: ${}", statement.closing_balance);
    /// # Ok::<(), stateset_embedded::CommerceError>(())
    /// ```
    pub fn generate_statement(
        &self,
        request: GenerateStatementRequest,
    ) -> Result<CustomerStatement> {
        self.db.accounts_receivable().generate_statement(request)
    }

    // ========================================================================
    // Analytics
    // ========================================================================

    /// Get total outstanding receivables.
    pub fn get_total_outstanding(&self) -> Result<Decimal> {
        self.db.accounts_receivable().get_total_outstanding()
    }

    /// Calculate Days Sales Outstanding (DSO).
    pub fn get_dso(&self, days: i32) -> Result<Decimal> {
        self.db.accounts_receivable().get_dso(days)
    }

    /// Get average days to pay for a customer.
    pub fn get_average_days_to_pay(&self, customer_id: Uuid) -> Result<Option<i32>> {
        self.db.accounts_receivable().get_average_days_to_pay(customer_id)
    }

    /// Get AR summary for multiple customers.
    pub fn get_customers_batch(&self, ids: Vec<Uuid>) -> Result<Vec<CustomerArSummary>> {
        self.db.accounts_receivable().get_customers_batch(ids)
    }
}

File: cli/src/commands/accounts-receivable.js
/**
 * Accounts Receivable Commands Module
 */

function parseDays(value, usage) {
  const parsed = Number.parseInt(value, 10);
  if (!Number.isInteger(parsed) || parsed < 1) throw new Error(usage);
  return parsed;
}

function parseAmount(value, usage) {
  const parsed = Number.parseFloat(value);
  if (!Number.isFinite(parsed) || parsed <= 0) throw new Error(usage);
  return parsed;
}

export async function execute(action, args, { commerce, output, jsonOutput }) {
  switch (action) {
    case 'aging': {
      const summary = await commerce.accountsReceivable.getAgingSummary();
      return jsonOutput
        ? summary
        : {
            summary,
            formatted:
              `Accounts receivable aging\n` +
              `${'-'.repeat(30)}\n` +
              `Current:      ${summary.current ?? 'N/A'}\n` +
              `1-30 days:    ${summary.days1to30 ?? 'N/A'}\n` +
              `31-60 days:   ${summary.days31to60 ?? 'N/A'}\n` +
              `61-90 days:   ${summary.days61to90 ?? 'N/A'}\n` +
              `90+ days:     ${summary.days90plus ?? 'N/A'}`,
          };
    }

    case 'outstanding': {
      const totalOutstanding = await commerce.accountsReceivable.getTotalOutstanding();
      return jsonOutput
        ? { totalOutstanding }
        : { formatted: `Accounts receivable total outstanding: ${totalOutstanding}` };
    }

    case 'dso': {
      const days = parseDays(args[0] || '30', 'Usage: accounts-receivable dso [days]');
      const dso = await commerce.accountsReceivable.getDso(days);
      return jsonOutput ? { days, dso } : { formatted: `DSO over ${days} days: ${dso}` };
    }

    case 'credit-memos': {
      const creditMemos = await commerce.accountsReceivable.listCreditMemos();
      return formatCreditMemos(creditMemos, { output, jsonOutput });
    }

    case 'credit-memo': {
      const creditMemoId = args[0];
      if (!creditMemoId) throw new Error('Usage: accounts-receivable credit-memo <creditMemoId>');
      const creditMemo = await commerce.accountsReceivable.getCreditMemo(creditMemoId);
      if (!creditMemo) throw new Error(`Credit memo not found: ${creditMemoId}`);
      return formatCreditMemo(creditMemo, { jsonOutput });
    }

    case 'create-credit-memo': {
      const [customerId, amountRaw, reason, originalInvoiceId, ...noteParts] = args;
      if (!customerId || !amountRaw || !reason) {
        throw new Error(
          'Usage: accounts-receivable create-credit-memo <customerId> <amount> <reason> [originalInvoiceId] [notes]',
        );
      }
      const creditMemo = await commerce.accountsReceivable.createCreditMemo({
        customerId,
        amount: parseAmount(
          amountRaw,
          'Usage: accounts-receivable create-credit-memo <customerId> <amount> <reason> [originalInvoiceId] [notes]',
        ),
        reason,
        originalInvoiceId: originalInvoiceId || undefined,
        notes: noteParts.join(' ') || undefined,
      });
      return {
        creditMemo,
        formatted: `Created credit memo ${creditMemo.id}`,
      };
    }

    case 'void-credit-memo': {
      const creditMemoId = args[0];
      if (!creditMemoId)
        throw new Error('Usage: accounts-receivable void-credit-memo <creditMemoId>');
      const creditMemo = await commerce.accountsReceivable.voidCreditMemo(creditMemoId);
      return { creditMemo, formatted: `Voided credit memo ${creditMemo.id}` };
    }

    case 'unapplied': {
      const customerId = args[0];
      if (!customerId) throw new Error('Usage: accounts-receivable unapplied <customerId>');
      const creditMemos = await commerce.accountsReceivable.getUnappliedCredits(customerId);
      return formatCreditMemos(creditMemos, { output, jsonOutput });
    }

    default:
      throw new Error(
        `Unknown action: accounts-receivable ${action}\n\n` +
          'Available actions:\n' +
          '  aging                                                                  Get AR aging summary\n' +
          '  outstanding                                                            Get AR total outstanding\n' +
          '  dso [days]                                                             Get days sales outstanding\n' +
          '  credit-memos                                                           List credit memos\n' +
          '  credit-memo <creditMemoId>                                             Get credit memo\n' +
          '  create-credit-memo <customerId> <amount> <reason> [originalInvoiceId] [notes]\n' +
          '  void-credit-memo <creditMemoId>                                        Void credit memo\n' +
          '  unapplied <customerId>                                                 List unapplied credits',
      );
  }
}

function formatCreditMemos(creditMemos, { output, jsonOutput }) {
  if (jsonOutput) return creditMemos;
  if (creditMemos.length === 0) return { formatted: 'No credit memos found.' };
  const formatted = output.table(creditMemos, [
    { key: 'id', header: 'ID' },
    { key: 'customerId', header: 'Customer' },
    { key: 'status', header: 'Status' },
    { key: 'amount', header: 'Amount', align: 'right' },
    { key: 'reason', header: 'Reason' },
  ]);
  return { creditMemos, formatted };
}

function formatCreditMemo(creditMemo, { jsonOutput }) {
  if (jsonOutput) return creditMemo;
  return {
    creditMemo,
    formatted:
      `Credit memo: ${creditMemo.id}\n` +
      `${'-'.repeat(34)}\n` +
      `Customer:       ${creditMemo.customerId}\n` +
      `Amount:         ${creditMemo.amount}\n` +
      `Status:         ${creditMemo.status}\n` +
      `Reason:         ${creditMemo.reason}\n` +
      `Invoice:        ${creditMemo.originalInvoiceId || 'N/A'}`,
  };
}

export const metadata = {
  name: 'accounts-receivable',
  aliases: ['ar', 'credit-memos'],
  description: 'Accounts receivable and credit memo commands',
  actions: {
    aging: { description: 'Get AR aging summary', args: [] },
    outstanding: { description: 'Get total outstanding', args: [] },
    dso: { description: 'Get days sales outstanding', args: ['[days]'] },
    'credit-memos': { description: 'List credit memos', args: [] },
    'credit-memo': { description: 'Get credit memo', args: ['<creditMemoId>'] },
    'create-credit-memo': {
      description: 'Create credit memo',
      args: ['<customerId>', '<amount>', '<reason>', '[originalInvoiceId]', '[notes]'],
    },
    'void-credit-memo': { description: 'Void credit memo', args: ['<creditMemoId>'] },
    unapplied: { description: 'List unapplied credits', args: ['<customerId>'] },
  },
};

export default { execute, metadata };
